### PR TITLE
AGENT-397: Improved logging for investigating skipped bytes

### PR DESCRIPTION
**Pull request by yanscalyr**
Still needs tests, and there is some more manual testing I need to do.

### CODE REVIEW

File: scalyr_agent/agent_main.py [hunk 1]
```diff
@@ -1658,6 +1749,59 @@ def __calculate_overall_stats(self, base_overall_stats):
...
+                    self.__last_total_bytes_skipped
+                    + self.__last_total_bytes_copied
+                    + self.__last_total_bytes_pending
+                )
+            ) / self.__config.copying_manager_stats_log_interval
+            if result.total_bytes_skipped > self.__last_total_bytes_skipped:
+                rate_limited_time = (
+                    result.total_rate_limited_time - self.__last_total_rate_limited_time
+                ) / self.__config.copying_manager_stats_log_interval
+                if self.__config.parsed_max_send_rate_enforcement:
```
Comment (Kami) [inline scalyr_agent/agent_main.py:1772]: Let's add tests for those two warnings.

Will likely need to mock a couple of things, but it should be mostly straight forward if we test this method directly.

File: scalyr_agent/agent_main.py [hunk 2]
```diff
@@ -884,6 +888,8 @@ def __run(self, controller):
         last_overall_stats_report_time = time.time()
         # We only emit the bandwidth stats once every minute.  Track when we last reported it.
         last_bw_stats_report_time = last_overall_stats_report_time
+        # We only emit the copying_manager stats once every 5 minutes.  Track when we last reported it.
+        last_copy_manager_stats_report_time = last_overall_stats_report_time
```
Comment (czerwingithub) [inline scalyr_agent/agent_main.py:892]: To improve readability here, let's declare a variable called `current_time` and set `time.time()` and use it in all of the above statements.  It looks a little odd when reading this line without that context.

File: scalyr_agent/agent_main.py [hunk 3]
```diff
@@ -1563,7 +1594,36 @@ def __log_bandwidth_stats(self, overall_stats):
             )
         )
 
-    def __calculate_overall_stats(self, base_overall_stats):
+    def __log_copy_manager_stats(self, overall_stats):
+        """Logs the copy_manager_status message that we periodically write to the agent log to give copying manager
+        stats.
+
+        This includes such metrics as the amount of times through the main look and time spent in various sections.
```
Comment (czerwingithub) [inline scalyr_agent/agent_main.py:1601]: /look/loop/

File: scalyr_agent/agent_main.py [hunk 4]
```diff
@@ -1658,6 +1749,59 @@ def __calculate_overall_stats(self, base_overall_stats):
             result.rss_size,
         ) = self.__controller.get_usage_info()
 
+        if copy_manager_warnings:
+            result.avg_bytes_copied_rate = (
+                result.total_bytes_copied - self.__last_total_bytes_copied
+            ) / self.__config.copying_manager_stats_log_interval
+            result.avg_bytes_produced_rate = (
```
Comment (czerwingithub) [inline scalyr_agent/agent_main.py:1756]: You could define two variables here called `total_bytes_produced` and `last_total_bytes_produced`.  It'd probably make the code a little more readable.  Alternatively, add a comment here that says you compute total produced by adding bytes_skipped, bytes_copied, and bytes_pending.

File: scalyr_agent/configuration.py [hunk 5]
```diff
@@ -2330,6 +2338,13 @@ def __verify_main_config_and_apply_defaults(
         self.__verify_or_set_optional_bool(
             config, "disable_leak_bandwidth_stats", False, description, apply_defaults
         )
+        self.__verify_or_set_optional_bool(
+            config,
+            "disable_leak_copy_manager_stats",
```
Comment (czerwingithub) [inline scalyr_agent/configuration.py:2343]: I don't think you mean `leak` here.

Comment (yanscalyr) [reply scalyr_agent/configuration.py:2343]: I went with how the other 2 status message disable flags were defined, I have no idea why they have "leak" there actually. I'm fine removing it.

Comment (czerwingithub) [reply scalyr_agent/configuration.py:2343]: Those were status messages put in to help look specifically for memory leak issues.

File: scalyr_agent/agent_main.py (hunk 2 above)
Comment (yanscalyr) [reply scalyr_agent/agent_main.py:892]: We have a `self.__start_time` set a bit above this, I think that would make sense to use here. Naming wise it seems pretty clear.

File: scalyr_agent/copying_manager.py [hunk 6]
```diff
@@ -800,7 +812,16 @@ def run(self):
                                 pipeline_time = 0.0
 
                             # Now block for the response.
+                            blocking_response_time_start = time.time()
                             (result, bytes_sent, full_response) = get_response()
+                            blocking_response_time_end = time.time()
```
[the remark this answers is not in the document]
Comment (yanscalyr) [reply scalyr_agent/copying_manager.py:817]: The `total_request_time` you mean? The notion doc does mention that this one will overlap with other timers. Should we exclude whatever time is spent in the send methods for this, or just avoid the compression specifically?

[the remark this answers is not in the document]
Comment (czerwingithub) [reply scalyr_agent/copying_manager.py:817]: Sorry, I meant removing it from `total_blocking_response_time`.  I'd like to try to have the `blocking_response_time` just be the time we are blocking on the server response.  How hard would that be to do?

**Comment by yanscalyr** [comment]
I don't believe `get_response` has any compression happening in it. Is there something specific you say that makes you think it has the compression time included?

**Comment by Kami** [comment]
@yanscalyr I believe it eventually calls this code - https://github.com/scalyr/scalyr-agent-2/blob/f96346854742f7e565d098549d5b30ba8cdd6810/scalyr_agent/scalyr_client.py#L653 which also prepares request, compresses it and sends it.

**Comment by yanscalyr** [comment]
That part of the code is reached after calling `send_events` here: https://github.com/scalyr/scalyr-agent-2/blob/f96346854742f7e565d098549d5b30ba8cdd6810/scalyr_agent/copying_manager.py#L785
The method we call in the blocking_time section is the callback that is returned by that call that did the compression, it only reads the response.

**Comment by czerwingithub** [comment]
> That part of the code is reached after calling `send_events` here:
> 
> https://github.com/scalyr/scalyr-agent-2/blob/f96346854742f7e565d098549d5b30ba8cdd6810/scalyr_agent/copying_manager.py#L785
> 
> 
> The method we call in the blocking_time section is the callback that is returned by that call that did the compression, it only reads the response.

Ok, thanks for verifying that.  Sgtm.

File: scalyr_agent/agent_main.py (hunk 1 above)
Comment (Kami) [reply scalyr_agent/agent_main.py:1775]: And eventually (post release is fine) we should also have some end tests which would cover that in combination with the new ``max_send_rate_enforcement`` configuration option.

We now have end to end tests in place so it shouldn't be too painful to add a test case for that scenario.

**Pull request merged**